x-pack/legacy/plugins/reporting/server/plugin.ts

```diff
@@ -49,52 +77,111 @@ export interface LegacySetup {
+    const { xpack_main: xpackMainPlugin, reporting } = __LEGACY.plugins;
+
+    const browserDriverFactory = await createBrowserDriverFactory(__LEGACY, this.logger); // required for validations :(
+    runValidations(__LEGACY, elasticsearch, browserDriverFactory, this.logger); // this must run early, as it sets up config defaults
 
-    let isCollectorReady = false;
+    // LEGACY
+    mirrorPluginStatus(xpackMainPlugin, reporting);
+    const checkLicense = checkLicenseFactory(this.exportTypesRegistry);
+    (xpackMainPlugin as any).status.once('green', () => {
+      // Register a function that is called whenever the xpack info changes,
+      // to re-compute the license check results for this plugin
+      xpackMainPlugin.info.feature(PLUGIN_ID).registerLicenseCheckResultsGenerator(checkLicense);
+    });
+    // Reporting routes
+    registerRoutes(this, __LEGACY, plugins, this.logger);
+    // END LEGACY
 
     // Register a function with server to manage the collection of usage stats
     registerReportingUsageCollector(
-      usageCollection,
       __LEGACY,
-      () => isCollectorReady,
-      exportTypesRegistry
+      usageCollection,
+      this.pluginStart$,
+      this.exportTypesRegistry
     );
 
-    const logger = new LevelLogger(this.context.logger.get('reporting'));
-    const browserDriverFactory = await createBrowserDriverFactory(__LEGACY, logger);
+    this.pluginSetup$.next({ browserDriverFactory });
+    return { browserDriverFactory };
+  }
 
-    logConfiguration(__LEGACY, logger);
-    runValidations(__LEGACY, elasticsearch, logger, browserDriverFactory);
+  public async start(core: CoreStart, plugins: ReportingStartDeps) {
+    const { elasticsearch, __LEGACY } = plugins;
 
-    const { xpack_main: xpackMainPlugin, reporting } = __LEGACY.plugins;
-    mirrorPluginStatus(xpackMainPlugin, reporting);
+    const esqueue = await createQueueFactory(this, __LEGACY, elasticsearch, this.logger);
+    const enqueueJob = enqueueJobFactory(this, __LEGACY, elasticsearch, this.logger);
 
-    const checkLicense = checkLicenseFactory(exportTypesRegistry);
+    this.pluginStartDeps = {
+      savedObjects: core.savedObjects,
+      uiSettings: core.uiSettings,
+      esqueue,
+      enqueueJob,
+    };
 
-    (xpackMainPlugin as any).status.once('green', () => {
-      // Register a function that is called whenever the xpack info changes,
-      // to re-compute the license check results for this plugin
-      xpackMainPlugin.info.feature(PLUGIN_ID).registerLicenseCheckResultsGenerator(checkLicense);
-    });
+    this.pluginStart$.next(this.pluginStartDeps);
 
-    // Post initialization of the above code, the collector is now ready to fetch its data
-    isCollectorReady = true;
+    setFieldFormats(plugins.data.fieldFormats);
+    logConfiguration(__LEGACY, this.logger);
 
-    // Reporting routes
-    registerRoutes(__LEGACY, plugins, exportTypesRegistry, browserDriverFactory, logger);
+    return this.pluginStartDeps;
```

Comment:
This is supposed to be the plugin's public contract. You are currently returning an internal contract, which doesn't really make any sense of any plugins that may want to consume your API (even if I get that `reporting` is more of a 'end of chain' plugin, this still doesn't respect the core conventions)
```typescript
this.pluginStartDeps = {
      savedObjects: core.savedObjects,
      uiSettings: core.uiSettings,
      esqueue,
      enqueueJob,
    };
```

My question is: I see you are using that internally (`this.pluginStart$.next(this.pluginStartDeps);`), but do you use anywhere this contract from the actual plugin's `start` return? 

- if no, I would stick to returning an empty contract (`{}`)
- If you do need to access this from outside of the shim for any reason, I would wrap it inside an `__internal` property to make it clear that this a temporary measure during migration

This follows my previous comment, I think your `pluginStartDeps` should be more named such as `reportingInternalStartDeps` or something like that, as it is supposed to be only consumed from inside your plugin.

Reply:
It makes senses to use a new types for `Internal*` instead of hooking into the types that were conveniently there :). Since I chose to use a type conveniently in scope, that forced me to return it. I get the point that it should only public expose internals as-needed by outside consumers

> if no, I would stick to returning an empty contract ({})

👍